docs/docs/vm-fundamentals/hints.md

```diff
@@ -2,6 +2,155 @@
+
+For example:
+ - **Assertion Checking**: Developers or testers use specific statements, known as assertions, to validate the expected   behavior of a program or system. Verification tools check whether these assertions fail during the analysis.
+    
+ - **Invariant Detection**: Hints may guide the tool to identify the uniformity of loop iteration or other program properties responsible for the accuracy of the contract. The tool then attempts to prove these properties using the execution trace.
+
+ - **Path Exploration**: Hints can help the tools explore specific paths or branches of the program that are considered fit for verification. It helps in focusing the analysis on relevant parts of the code.
+   
+
+5. **Feedback Loop**: As the verification tool analyzes the execution trace and attempts to prove properties guided by the hints, it may encounter challenges or limitations. In such cases, feedback mechanisms allow developers to refine their hints or modify the code to improve the chances of successful verification.
+
+### WHY HINTS RUN ON SEQUENCERS/ PROVERS AND NOT VERIFIERS
+
+The verifier in Cairo is responsible for ensuring that a program attaches safety and accuracy features. It checks that the program satisfies limitations such as memory safety, type safety, and functional correctness.
+
+Sequencers in Cairo are responsible for organizing the execution of transactions, managing instruction ordering and dependencies, exploiting multiple processors, and ensuring the integrity of the program's state throughout the execution process.   
+     
+Provers attempt to mathematically prove the correctness of a program concerning a given specification. The prover uses these hints to access the program's behavior and verify that it satisfies specified properties and constraints. In this context, hints could guide the prover towards particular properties or parts of the codebase that are relevant to the proof.
+
+During the proof generation process, the prover uses hints to generate additional constraints or assumptions about the program's behavior, which makes part of the proof sent to the verifier for verification. In the Cairo VM, hints are sorted out during different stages of the compilation and execution process by the sequencer and the prover. However, they are not directly involved in the verification stage. Hints are primarily for performance optimization.
+    
+Sequencers and provers are tools used in formal verification, a process of accurately proving the correctness of programs.
+Applying hints in the Cario virtual machine leverages the execution trace and uses it to guide the verification process. Integration of hints with the Cairo VM's execution trace and verification tools enables developers to guide and customize the verification process, improving the efficiency and effectiveness of formal verification for smart contracts.
```

Comment:
Could you point out the resource you used for this part? I'm not sure hints are related to formal verification techniques.

Also I would like to add in this documentation some information regarding the way we implement hints. You can use the content from https://github.com/NethermindEth/cairo-vm-go/blob/main/pkg/hintrunner/zero/README.md as reference.